Slim/Networking/SimpleAsyncWS.pm

```diff
@@ -0,0 +1,396 @@
+package Slim::Networking::SimpleAsyncWS;
+
+# Lyrion Music Server Copyright 2024 Lyrion Community.
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License,
+# version 2.
+
+# This class provides a non-blocking WebSockets client connection from Lyrion Music Server.
+
+# This class is intended for plugins and other code needing simply to
+# handle a persistent websockets connection.  If you have more complex
+# needs consider writing a fuller implementation.
+
+# more documentation at end of file.
+
+use strict;
+
+use IO::Socket;
+use IO::Socket::SSL;
+use IO::Select;
+use Protocol::WebSocket::Client;
+
+use Data::Dumper;
+
+use Slim::Utils::Log;
+use Slim::Utils::Prefs;
+
+my $log = logger('network.asyncws');
+
+sub new {
+	my ( $class, $url, $cbConnected, $cbConnectFailed) = @_;
+
+	my $self = {
+		client     => 0,
+		tcp_socket   => 0,
+		socket_open  => 0,
+		continue_listening => 0,
+		cb_Read  => 0,
+		cb_Read_Failed => 0,
+	};
+
+	bless $self, $class;
+
+	$self->_connect( $url, $cbConnected, $cbConnectFailed );
+
+	return $self;
+}
+
+
+sub close {
+	my ($self) = @_;
+
+	main::INFOLOG && $log->is_info && $log->info("Close web socket connect with status: " . $self->{tcp_socket}->connected() );
+
+	$self->{continue_listening} = 0;
+	$self->{client}->disconnect;
+	$self->{tcp_socket}->close if $self->{socket_open};
+	$self->{socket_open} = 0;
+
+	return;
+}
+
+
+sub _connect {
+	my ( $self, $url, $cbConnected, $cbConnectFailed ) = @_;
+
+	main::DEBUGLOG && $log->is_debug && $log->debug("Connecting to webSocket $url");
+
+	my ($proto, $host, $port, $path);
+	if ($url =~ m/^(?:(?<proto>ws|wss):\/\/)?(?<host>[^\/:]+)(?::(?<port>\d+))?(?<path>\/.*)?$/){
+		$host = $+{host};
```

Comment:
I'm not familiar with that `$+{variable}` construct... could you please tell me?

Reply:
Haha, I would have lifted that whole web socket url code from a Protocol::Websocket library usage sample when I put it together some time ago!.  It may have even been lifted from Anynet::Websocket.    I should really go back and check where I got some of the sample code from and credit at the top of the file.
Anyway,  I believe that gets just the host from the named capture group in the regex above.

Reply:
I've now removed this bespoke url parseing by adding URI::ws to CPAN.  I can now use URI->new to parse the ws url.